terraform/jenkins/README.md

```diff
@@ -0,0 +1,185 @@
+(re-)configure list of trusted public keys with something like cloud-init, as
+it's baked into the /etc/nix/nix.conf that's generated by the filesystem.
+
+This causes our environment-agnostic images to still be specific to a set of
+public keys.
+
+We can probably fix this by extending `nix.conf` on bootup, and pointing
+`nix-daemon.service` to the extended config file.
+
+### Jenkins configuration and authentication
+Currently, Jenkins is configured purely imperatively, using its state volume for
+pipeline config. It also has no user setup (yet), but logs an admin password
+that's supposed to be used for login.
+
+We should configure some Jenkins pipelines, probably via cloud-init (so we don't
+need to bake new images all the time), configure SSO for login and properly
+expose this via a domain and HTTPS.
+
+### More dynamic nix builder registration
+Nix reads the list of available builders from `/etc/nix/machines`.
+
+This file is assembled with terraform, creating a strong coupling between this
+file and all builder machines.
+
+It is written to disk by cloud-init once on startup, and the same list of
+builders is scanned for ssh host keys once on bootup.
+
+This makes registering new builders quite a churn:
+
+ - It requires an update of the cloud-init userdata and VM recreation to update
+   that list (or manual tinkering over ssh), stopping all builds.
+ - It (currently) doesn't allow explicitly specifying host keys.
+   While we don't know until the target machine has booted up, we might know for
+   longer-running builders.
+ - Redeploying a builder causes its host key to change, requiring the known host
+   keys to be updated.
+ - managing ssh private keys in general is annoying, and there's little reason
+   to allow ingoing ssh.
+
+There should be a more dynamic "agent-based" registration process.
+
+Ideally, the builders could register themselves with the controller, advertise
+their capabilities (number of cores, architectures), and keep the connection alive.
+
+`/etc/nix/machines` on the controller could then provide a "live view" into
+the currently connected builders, and give a unix domain socket (using `unix://
+` URL) that'll connect to a `nix daemon --stdio` on the other side of the
+connection.
+
+Authentication TBD. At least in Azure, this could use Machine Identity.
+
+### Offload signing
+We have the Nix binary cache signing key as a literal file (only accessible for
+root), and let `nix copy` take care of creating signatures with it.
+
+Obviously this is very bad, as leaving the private key on the host makes it at
+least possible to steal the private key.
+
+It seems Azure Key Vault supports dealing with ed25519 key material, and doing
+signatures. Rather than keeping the key as a file on disk.
+
+A much cleaner solution would be to allow offloading the signing mechanism to
+Azure Key Vault, or "external signers in general".
+
+In terms of complexity, there's already
```

Comment:
Even simpler solution: Generate hash & send REST API to KeyVault.

Reply:
Nix uses a fingerprint of a NAR and requires a signature for that. There's currently no direct mechanism inside Nix to retrieve this [fingerprint[(https://docs.tvix.dev/rust/nix_compat/narinfo/fn.fingerprint.html)